### PR TITLE
Multisite authentication efficiency

### PR DESCRIPTION
**Pull request by adam-vessey**
**JIRA Ticket**: [ISLANDORA-1940](https://jira.duraspace.org/browse/ISLANDORA-1940)

# What does this Pull Request do?

Introduces a JAAS LoginModule with provisions for handling additional context for authentication, in order select the site against which to authenticate, moving the lookup from linear to constant time.

Also, has a somewhat unfortunate number of whitespace changes, due to a combination of tabs and spaces having been used for indentation (now, uses spaces, in any files which were changed).

# What's new?

This PR introduces a few new/modified moving parts:

* JAAS implementation to acquire additional context (the header)
* LoginModule implementation to _use_ the additional context
* Connection configuration in the LoginModule (to inform the use of the context)

# How should this be tested?

1. Setup multisites pointing at a single Fedora.
2. For the Drupal filter, follow the [multisite optimization configuration instructions](https://github.com/adam-vessey/islandora_drupal_filter/blob/dbe5e3b58b089bc1833325bfaa35563995c9f413/docs/multisite-optimization.md).

# Additional Notes:

* **Does this change require documentation to be updated?** Possibly; some in README (and other file) already.
* **Does this change add any new dependencies?** Explicit ones, no. To make use of the new configuration, it would be necessary to inject the required header into requests in some manner.
* **Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)?** No.
* **Could this change impact execution of existing code?** Being used as is (in current configurations), no. When moving to use the multisite-optimized configuration without the fallback to the present configuration, any invocations against Fedora directly using a username and password hash from the Drupal database name would have to follow suit, providing the header identifying the site from which the credentials originate.

# Interested parties:

@Islandora/7-x-1-x-committers ... No component manager?

### CODE REVIEW

**Comment by jonathangreen** [comment]
Just taking a look over this. I just reworked how we do authentication against Fedora in CLAW using headers here: https://github.com/Islandora-CLAW/CLAW/issues/520

@adam-vessey would you be able to make a commit that fixes the whitespace issues, then rebase this on top of it, so that I can look at a diff without all the whitespace changes? Be much easier to read and review.

**Comment by adam-vessey** [comment]
@jonathangreen: Do you mean something like [this](https://github.com/Islandora/islandora_drupal_filter/compare/master...adam-vessey:ISLANDORA-1940-space-rebased), so you can see the [comparison without the spaces](https://github.com/Islandora/islandora_drupal_filter/compare/8df0b2d...adam-vessey:ISLANDORA-1940-space-rebased)? (Messed up a couple of chunks, will fix).

File: src/test/resources/filter-drupal-multisite.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<FilterDrupal_Connection>
+	<connection key="test_key" server="${db.server}" port="${db.port}" dbname="${db.name}" user="${db.user}" password="${db.pass}" jdbcDriverClass="${db.class}" jdbcURLProtocol="${db.scheme}">
+	    <sql>
+	       SELECT DISTINCT u.uid AS userid, u.name AS Name, u.pass AS Pass,r.name AS Role
+           FROM (users u LEFT JOIN users_roles ON u.uid=users_roles.uid) LEFT JOIN role r ON r.rid=users_roles.rid
+           WHERE u.name=? AND u.pass=?; 
+	    </sql>
+    </connection>
+</FilterDrupal_Connection>
```
Comment (willtp87) [inline src/test/resources/filter-drupal-multisite.xml]: New line for git please?

File: src/main/java/ca/upei/roblib/fedora/servletfilter/jaas/MissingCredsException.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+public class MissingCredsException extends UnsupportedCallbackException {
+
+    /**
+     *
+     */
+    private static final long serialVersionUID = 5743041997336340500L;
+
+    public MissingCredsException(Callback callback) {
+        super(callback);
+        // TODO Auto-generated constructor stub
```
Comment (willtp87) [inline src/main/java/ca/upei/roblib/fedora/servletfilter/jaas/MissingCredsException.java]: Eclipse games? probably shouldn't be around?

**Comment by willtp87** [comment]
I'm sorry but we will need to close this pull because it does not comply with the CLA as it contains some work we do not posses the IP for.  We will be making the work available in another fashion and will update this thread when that is complete.

**Comment by manez** [comment]
Thanks for your diligence, @willtp87

**Comment by DiegoPino** [comment]
@willtp87 do you want to talk about that during our next release/committers
 call?

**Pull request closed without merging**